### ISSUE TITLE
Question about analog SGD

### ISSUE THREAD

**Issue by Winfriedvdd** [IBM/aihwkit#338]
Hi! I have a few questions about the analog SGD that is implemented, and what alternatives can be used for training.

If my understanding is correct, the analog SGD is meant to mimic training on the actual memristor devices.
However, I am interested in the situation where one would first train a network under optimal conditions, and only then implement the weights on analog layers.
With this I do not mean training in digital or without noise, because such conditions would not be "optimal" in the sense that the training would not be hardware-aware.

An example of an issue that may arise: 
In ideal noise-aware training, e.g., using a reparameterization trick, one would use the same random "draw" of the weights from their distributions to compute both the forward and the backward pass, if estimating the gradient of the loss function. (Instead of randomly drawing weights for the forward pass, and then again for the backward pass.)
I can see how this would be a bit of a problem when training on the actual hardware, but what are the options for how this is handled in the simulation?
Are the weights used in the forward pass generated using a different (i.e., newly drawn) noise source compared to the weights used in the backward pass?

Noticing that there are many buttons and options to play with, considering rpu_configurations and their input/output/forward/backward parameters etc., I was wondering what setup would be best to use if I want to train in the most optimal setting. In other words, how would you train when not trying to mimic suboptimal training on the hardware, but when you also don't want to train with full precision/in digital either. (Rather something in between)

Are there for example already some noise sources that I should probably turn off because they are only related to specific sub-optimal "training on hardware" conditions? Or are there altogether specific preset rpu_configurations that are interesting for this purpose?

I noticed that I can also use standard pytorch SGD on analog layers, how is this dealt with under the hood? Does it revert to AnalogSGD automatically, recognizing that the layers are in fact analog? Or is something else happening when using standard SGD?

Sorry for the long question!

**Comment by maljoras** [comment]
Hi @Winfriedvdd ,
many thanks for your question! 

Indeed the toolkit provides analog training (where the update is done on the analog array directly) and also "software" training, that we call "hardware-aware" training, for DNNs that are to be programmed onto an analog array for inference purposes only. For the latter, we use the `InferenceRPUConfig`. This `RPUConfig` is dedicated for analog inference only chips, and will do hardware-aware training if used during training in AIHWKIT. You can specify the non-idealities and noise-sources of the forward pass only, since the backward pass and update is perfect in this case. The parameters specified in the "forward" field are indeed noises that are *only* applied during the forward pass (each mat-vec drawn anew even if a tile is used multiple times per mini-batch) and are not included in the backward pass, which is ideal and noise free. 

However, we also support the same "weight noise" during forward *and* backward: for that we have the `modifier` field in the rpu_config. This weight noise will be applied once per minibatch and the weight matrix is then kept fixed for the duration of one minibatch in forward and backward. The update is done on the non-disturbed weight matrix.  See [here](https://aihwkit.readthedocs.io/en/latest/api/aihwkit.simulator.configs.utils.html?highlight=modifier#aihwkit.simulator.configs.utils.WeightModifierParameter) for parameters related to such weight  modifications during HWA training. For instance, as in [example 5](https://github.com/IBM/aihwkit/blob/master/examples/05_simple_layer_hardware_aware.py#L46), one can do 
```python

# Imports from aihwkit.
from aihwkit.nn import AnalogLinear
from aihwkit.optim import AnalogSGD
from aihwkit.simulator.configs import InferenceRPUConfig
from aihwkit.simulator.configs.utils import (
    WeightNoiseType, WeightClipType, WeightModifierType)
from aihwkit.inference import PCMLikeNoiseModel, GlobalDriftCompensation
from aihwkit.simulator.rpu_base import cuda

# Define a single-layer network, using inference/hardware-aware training tile
rpu_config = InferenceRPUConfig()
rpu_config.forward.out_res = -1.  # Turn off (output) ADC discretization.
rpu_config.forward.w_noise_type = WeightNoiseType.ADDITIVE_CONSTANT
rpu_config.forward.w_noise = 0.02  # Short-term w-noise.

rpu_config.clip.type = WeightClipType.FIXED_VALUE
rpu_config.clip.fixed_value = 1.0
rpu_config.modifier.type = WeightModifierType.ADD_NORMAL  # Fwd/bwd weight noise.
rpu_config.modifier.std_dev = 0.01

# Inference noise model.
rpu_config.noise_model = PCMLikeNoiseModel(g_max=25.0)

# drift compensation
rpu_config.drift_compensation = GlobalDriftCompensation()

model = AnalogLinear(4, 2, bias=True,
                     rpu_config=rpu_config)
```
Note again that the `forward` setting are applied each mat-vec for foward only, wheras the `modifier` settings are weight noise to the weight in both forward and backward, updated once per mini-batch.  

In case you don't want to use any forward non-idealities, you can set `rpu_config.foward.is_perfect = True` in the above. The weight noise modifier would still be used in this case. 

We also now have a [notebook](https://github.com/IBM/aihwkit/blob/master/notebooks/examples/analog_training_LeNet5_hwa.ipynb) explaining hardware-aware training for inference with the toolkit

**Comment by Winfriedvdd** [comment]
Thanks for the quick and extensive response!
